pkg/scheduler/preemption/preemption.go

```diff
@@ -336,14 +337,42 @@ func workloadFits(wlReq cache.FlavorResourceQuantities, cq *cache.ClusterQueue,
 				cohortResRequestable = cq.Cohort.RequestableResources[flvQuotas.Name]
 			}
 			for rName, rReq := range flvReq {
+				flagBorrow, flagLend := false, false
 				limit := flvQuotas.Resources[rName].Nominal
+				if !allowBorrowing || cq.Cohort == nil {
+					if cqResUsage[rName]+rReq > limit {
+						return false
+					}
+				}
 
 				// No need to check whether cohort is nil here because when borrowingLimit
 				// is non nil, cohort is always non nil.
 				if allowBorrowing && flvQuotas.Resources[rName].BorrowingLimit != nil {
+					flagBorrow = true
 					limit += *flvQuotas.Resources[rName].BorrowingLimit
+					if cqResUsage[rName]+rReq > limit {
+						return false
+					}
 				}
 
+				// TODO: add validation for lendingLimit in cohort
+				if allowBorrowing && cq.Cohort != nil && features.Enabled(features.LendingLimit) {
+					if flvQuotas.Resources[rName].LendingLimit != nil {
+						flagLend = true
+						// If LendingLimit is not nil here, we need to recalculate cohort.RequestableResources,
+						// because we are borrowing from the cohort now, need to count our own nominal quota in.
+						if guarantee, exist := cq.GetGuaranteedQuota(flvQuotas.Name, rName); exist {
+							cohortResRequestable[rName] += guarantee
+						}
+					}
+				}
+				if flagBorrow && flagLend {
+					// if lendingLimit and borrowingLimit both set
+					limit = min(cohortResRequestable[rName], limit)
+				} else {
+					// if one of lendingLimit and borrowingLimit unset
+					limit = max(cohortResRequestable[rName], limit)
+				}
 				if cqResUsage[rName]+rReq > limit {
```

Comment:
Based on my understanding, what we need is only change L380:L82 to:
```
if cq.Cohort != nil {
    requestQuota :=  cohortResRequestable[rName]
    if lendingLimit != nil {
        requestQuota += guarantee
    }

    if cohortResUsage[rName]+rReq > requestQuota {
        return false
    }
}
```

Then we can make sure  that the borrowing resources will not exceed.